x-pack/plugins/security/server/lib/authentication/providers/oidc.js

```diff
@@ -0,0 +1,476 @@
+
+      this._options.log(['debug', 'security', 'oidc'], 'Request has been authenticated via refreshed token.');
+
+      return AuthenticationResult.succeeded(
+        user,
+        { accessToken: newAccessToken, refreshToken: newRefreshToken }
+      );
+    } catch (err) {
+      this._options.log(['debug', 'security', 'oidc'], `Failed to refresh elasticsearch access token: ${err.message}`);
+
+      // Reset `Authorization` header we've just set. We know for sure that it hasn't been defined before,
+      // otherwise it would have been used or completely rejected by the `authenticateViaHeader`.
+      // We can't just set `authorization` to `undefined` or `null`, we should remove this property
+      // entirely, otherwise `authorization` header without value will cause `callWithRequest` to fail if
+      // it's called with this request once again down the line (e.g. in the next authentication provider).
+      delete request.headers.authorization;
+
+      // There are at least two common cases when refresh token request can fail:
+      // 1. Refresh token is valid only for 24 hours and if it hasn't been used it expires.
+      //
+      // 2. Refresh token is one-time use token and if it has been used already, it is treated in the same way as
+      // expired token. Even though it's an edge case, there are several perfectly valid scenarios when it can
+      // happen. E.g. when several simultaneous AJAX request has been sent to Kibana, but Elasticsearch access token has
+      // expired already, so the first request that reaches Kibana uses refresh token to get a new elasticsearch access
+      // token, but the second concurrent request has no idea about that and tries to refresh access token as well. All
+      // ends well when first request refreshes access token and updates session cookie with fresh access/refresh token
+      // pair. But if user navigates to another page _before_ AJAX request (the one that triggered token refresh)
+      // responds with updated cookie, then user will have only that old cookie with expired access token and refresh
+      // token that has been used already.
+      //
+      // When user has neither valid access nor refresh token, the only way to resolve this issue is to re-initiate the
+      // OpenID Connect authentication by requesting a new authentication request to send to the OpenID Connect Provider
+      // and exchange it's forthcoming response for a new Elasticsearch access/refresh token pair. In case this is an
+      // AJAX request, we just reply with `400` and clear error message.
+      // There are two reasons for `400` and not `401`: Elasticsearch search responds with `400` so it seems logical
+      // to do the same on Kibana side and `401` would force user to logout and do full SLO if it's supported.
+      if (isInvalidRefreshTokenError(err)) {
+        if (canRedirectRequest(request)) {
+          console.log(
```

Comment:
nit: same comment as above, let's change this to `this._options.log`.

Reply:
uncleaned debugging:(